### PR TITLE
Feature/split client tools

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose

for better structural overview split the client-tools handling code into its own file.

- [x] :hammer: Refactoring/simplification

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -739,15 +739,9 @@ There are several major places where unittests live:
...
-| `js/client/modules/@arangodb/testutils/crash-utils.js`       | if something goes wrong, this contains the crash analysis tools                                                                   |
-| `js/client/modules/@arangodb/testutils/clusterstats.js`      | can be launched separately to monitor the cluster instances and their resource usage                                               |
-| `js/client/modules/@arangodb/testsuites/`                    | modules with testframework that control one set of tests each                                                                      |
-| `js/common/modules[/jsunity]/jsunity.js`                     | jsunity testing framework; invoked via jsunity.js next to the module                                                               |
-| `js/common/modules/@arangodb/mocha-runner.js`                | wrapper for running mocha tests in arangodb                                                                                        |
+
+                                          |
+see [js/client/modules/@arangodb/README.md] about implementaiton deatails of the framework.
```
Comment (mpoeter) [inline CONTRIBUTING.md:744]:
```suggestion
see [js/client/modules/@arangodb/README.md] about implementation details of the framework.
```

File: js/client/modules/@arangodb/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
+=== Arango DB test infrasturcture ===
+== Philosopy ==
+All the code structure and the test infrastructure works in a way to
+- have results in all cases
+- must not lock up, and terminate in a reasonable timeframe under all circumstances
+- waste as less host resources as possible ->
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:6]:
```suggestion
- waste as little host resources as possible -> 
```

File: js/client/modules/@arangodb/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
+=== Arango DB test infrasturcture ===
+== Philosopy ==
+All the code structure and the test infrastructure works in a way to
+- have results in all cases
+- must not lock up, and terminate in a reasonable timeframe under all circumstances
+- waste as less host resources as possible -> 
+  - detect lockups and abort tests (C++ / JS infrastructure)
+  - have a well established timeout and deadline mechanism 
+  - have a subprocess monitoring to abort as fast as possible if
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:9]: Sentence seems cut off - if what?

File: js/client/modules/@arangodb/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+| :----------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------- |
+| `testutils/testing.js`                                       | invoked via `unittest.js` handles module structure for `testsuites`.                                        |
+| `testutils/result-processing.js`                             | contains code to run various analyses on results and/or generate result representations for the user/ci     |
+| `testutils/test-utils.js`                                    | infrastructure for tests like filtering, bucketing, iterating                                               |
+| `testutils/process-utils.js`                                 | manage arango instances, start/stop/monitor SUT-processes                                                   |
+| `testutils/result-processing.js`                             | work with the result structures to produce reports, hit lists etc.                                          |
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:23]: result-processing.js is already described in line 20. Did you mean to write about a different file her (or there), or are both descriptions about the same file and should therefore be merged?

File: js/client/modules/@arangodb/README.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+| `testsuites/[go|php|java|js|driver|rta_makedata].js`         | subsidize the lifecycle method of `testRunner` for other instance and testrunner controls.                  |
+| `testsuites/[server_permissions|recovery[_cluster]].js`      | run several instances and run individual tests on them; vary parameters etc. of SUT                         |
+| `testsuites/chaos.js`                                        | launches several sub-arangoshs to cause randomness, enables failurepoints and so on.                        |
+| `js/common/modules[/jsunity]/jsunity.js`                     | jsunity testing framework; invoked via jsunity.js next to the module                                        |
+| `js/common/modules/@arangodb/mocha-runner.js`                | wrapper for running mocha tests in arangodb                                                                 |
+| `js/common/modules/@arangodb/mocha-runner.js`                | wrapper for running mocha tests in arangodb                                                                 |
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:37]:
```suggestion
| `js/common/modules/@arangodb/mocha-runner.js`                | wrapper for running mocha tests in arangodb                                                                 |
```

File: js/client/modules/@arangodb/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+- optional commandline parameters to be handled via `config`.
+- a list of files associated with the testsuites; This is used for the `auto` mechanism to associate a testfile with a suite.
+- `exports.setup` function to:
+  - register the suites directories containing integration testsuites with `allTestPaths` to to the testsuite.
+  - register the functions of the testsuites in `testFns`
+  - register the testsuite documentations with `fnDocs`
+
+== testsuite run integration ==
+The suite function. It should launch the SUT, identify & select the integration testsuites to run, shutdown the SUT and return the results.
+To achieve this it should lean on the various infrarstructure presented below.
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:64]:
```suggestion
To achieve this it should lean on the various infrastructure presented below.
```

File: js/client/modules/@arangodb/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+== result integration ==
+The result should be in a structure that can later be analyzed using `result-processing.js`.
+
+
+=== test infrastructure ===
+The test infrastructure should be used by testsuites to do most of the heavy lifting; in best case it can contain just out of few lines.
+== integration testsuites management ==
+A testsuite receives a list of testfiles which each one is a single `integration testsuite`. Its files located by `testing.js` from the `allTestPaths` directory list.
```
Comment (mpoeter) [inline js/client/modules/@arangodb/README.md:73]:
```suggestion
A testsuite receives a list of testfiles where each one is a single `integration testsuite`. Its files located by `testing.js` from the `allTestPaths` directory list.
```

File: js/client/modules/@arangodb/README.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
+=== Arango DB test infrasturcture ===
+== Philosopy ==
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:2]:
```suggestion
== Philosophy ==
```

File: js/client/modules/@arangodb/README.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+All the code structure and the test infrastructure works in a way to
+- have results in all (worst-) cases
+- must not lock up, and terminate in a reasonable timeframe under all circumstances
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:5]:
```suggestion
- not lock up, and terminate in a reasonable timeframe under all circumstances
```

File: js/client/modules/@arangodb/README.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+- waste as little host resources as possible -> 
+  - detect lockups and abort tests (C++ / JS infrastructure)
+  - have a well established timeout and deadline mechanism 
+  - have a subprocess monitoring to abort as fast as possible if spawned processes go away
+    (i.e. clusters may get hell slow once instances are missing)
+- collect as much data from the system as possible
+- compress the data to present the user with a most compact result
+- Timeout is for individual integration testsuites. Its 15 minutes by default.
+- Deadline is for the total load of tests.
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:14]:
```suggestion
Timeout is for individual integration testsuites. Its 15 minutes by default.
Deadline is for the total load of tests. 
```

File: js/client/modules/@arangodb/README.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+An important part of the `testRunner` is its API to register tests that ensure SUT-cleanness after individual integration testsuites.
+This way it is maintained, that individual integration testsuites don't leave things behind which may have influence on subsequent integration testsuites. 
+It consists of: 
+- `name` of the ensurer
+- `setUp` to be ran before the integration testsuite is launched to establish the "desired" dataset 
+- `runCheck` to be ran after the integration testsuite to check the SUT has the same state as before.
+Both `setUp` and `runCheck` can indicate the SUT is not fit for further tests, and abort the execution.
+
+The object manages its subprocesses in the `instanceManager` and uses it to manipulate processes.
+The API consists of several  default callback hooks that can be implemented dubbed `pre*()`/`post*()`.
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:91]:
```suggestion
The API consists of several default callback hooks that can be implemented dubbed `pre*()`/`post*()`.
```

File: js/client/modules/@arangodb/README.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+- `rta_makedata.js` - `rtaMakedataRunner` will implement its own `runOneTest` hook which a method that runs all phases of the test in one go; hence the phases and results don't match directly into the structure.
+- `chaos.js` - `chaosRunner` will launch the SUT, and run a set of arangoshs with parallel test executions on the SUT, and collect them in the end.
+- `[go|php|java|js|driver].js` will replace the `run` hook by launching the SUT on their own, and running  respective test-suites of these drivers, providing them with their instance, and convert their respective test-results back to `testing.js` own format.
+- `[server_permissions|recovery[_cluster]].js` will also replace the `run` function. It will launch one SUT per iteration and run its respective tests with it.
+
+== instance management ==
+An arangodb deployment may consist of several instances that are to be orchestrated in order to form the SUT. 
+Depending on the `options` the deployment mode is chosen by the `instanceManager`.
+It forms the `instance` objects of the respective `arangod`s to be launched and stopped.
+It knows the concetps of the agency and how to work with it.
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:122]:
```suggestion
It knows the concepts of the agency and how to work with it.
```

**Review by jsteemann — APPROVED** [review]
LGTM

File: js/client/modules/@arangodb/README.md [hunk 13]
```diff
@@ -1,8 +1,8 @@
-=== Arango DB test infrasturcture ===
-== Philosophy ==
+# Arango DB test infrasturcture
```
Comment (jsteemann) [inline js/client/modules/@arangodb/README.md:1]:
```suggestion
# Arango DB test infrastructure
```

**Pull request merged**